### ISSUE TITLE
Simple Seq2Seq Computes Incorrect Dev Log Loss

### ISSUE THREAD

**Issue by vidurj** [allenai/allennlp#1134]
When computing the log loss on the dev set, simple_seq2seq does not condition on gold tokens. This means that the loss reported on the dev set is not the negative log likelihood of the gold sequence.

**Comment by murphp15** [comment]
@vidurj 
Can I get more information on this issue, looking at the code I can't see an issue? 
I have looked at SimpleSeq2Seq and as far as I can see it does condition on the gold sequence? 
The method _get_loss takes in the logits and targets and computes the cross entropy. 

What am I missing? I will work on it as soon as I understand the issue.

**Comment by matt-gardner** [comment]
@vidurj, correct me if I'm wrong, but I believe the issue is in this block of code: https://github.com/allenai/allennlp/blob/b71ef434b310e9b1ba400b42c7334c794996b40e/allennlp/models/encoder_decoders/simple_seq2seq.py#L146-L155

This determines what to use as input to the decoder LSTM.  If we're training and not sampling from the model's output, we use the original targets, otherwise we use the model's output from the previous timestep.  What @vidurj wants is an option to always use the top branch during validation (or perhaps calculate that as a separate metric?  that'd be expensive, though, because you'd do two runs at validation time).

**Comment by murphp15** [comment]
Can he not set self._scheduled_sampling_ratio to be 0 during validation?

**Comment by matt-gardner** [comment]
And a new parameter, something like `compute_gold_dev_loss`, that goes in the constructor (and `from_params`).  Just adding another branch would make for spaghetti code - this can be simplified by adding a variable like `use_gold_targets` that gets set to `True` in some cases:

```python
# Not certain this logic is correct as-is; it should be close, though.
use_gold_targets = False
if self.training and all(...):
    use_gold_targets = True
elif timestep == 0:
    use_gold_targets = True  # We're assuming that the data always starts with the start token.
elif not self.training and self._compute_gold_dev_loss:
    use_gold_targets = True

if use_gold_targets:
    input_choices = targets[:, timestep]
else:
    input_choices = last_predictions
```

Though, having that flag passed to the constructor means that after training the model, you have to override the flag to actually make predictions...  @DeNeutoy, do you have any better ideas about how to handle this?

**Comment by vidurj** [comment]
FWIW, I fixed this for myself by conditioning on target_tokens at test time if they are provided i.e.
```
use_gold_targets = False
if self.training and all(...):
    use_gold_targets = True
elif timestep == 0:
    use_gold_targets = True  # We're assuming that the data always starts with the start token.
elif not self.training and target_tokens is not None:
    use_gold_targets = True

if use_gold_targets:
    input_choices = targets[:, timestep]
else:
    input_choices = last_predictions
```

**Comment by matt-gardner** [comment]
Closing this as stale, as we're not really working on the seq2seq model anymore at this point.

**Comment by brendan-ai2** [comment]
I'll make a quick PR for this and the other seq2seq issue. It's too nasty for us to leave lying around.

**Comment by brendan-ai2** [comment]
Fixed in https://github.com/allenai/allennlp/pull/1808.
